lib/ReorderBasicBlocksPass.cpp

```diff
@@ -41,21 +54,94 @@ namespace clspv {
 FunctionPass *createReorderBasicBlocksPass() {
   return new ReorderBasicBlocksPass();
 }
+} // namespace clspv
+
+void ReorderBasicBlocksPass::StructuredOrder(BasicBlock *block,
+                                             const LoopInfo &LI,
+                                             std::deque<BasicBlock *> *order,
+                                             DenseSet<BasicBlock *> *visited) {
+  if (!visited->insert(block).second)
+    return;
+
+  // Identify the merge and continue blocks for special treatment.
+  const auto *terminator = dyn_cast<TerminatorInst>(block->getTerminator());
+  BasicBlock *continue_block = nullptr;
+  BasicBlock *merge_block = nullptr;
+  if (LI.isLoopHeader(block)) {
+    Loop *loop = LI.getLoopFor(block);
+    merge_block = loop->getExitBlock();
+
+    if (loop->isLoopLatch(block)) {
+      continue_block = block;
+    } else {
+      DominatorTree &DT = getAnalysis<DominatorTreeWrapperPass>().getDomTree();
+      auto *header = loop->getHeader();
+      auto *latch = loop->getLoopLatch();
+      for (auto *bb : loop->blocks()) {
+        if (bb == header)
+          continue;
+
+        if (DT.dominates(bb, latch))
+          continue_block = bb;
+      }
+    }
+    assert(continue_block && merge_block && "Bad loop");
+  } else if (terminator->getNumSuccessors() > 1) {
+    bool has_back_edge = false;
+
+    for (unsigned i = 0; i < terminator->getNumSuccessors(); ++i) {
+      if (LI.isLoopHeader(terminator->getSuccessor(i)))
+        has_back_edge = true;
+    }
+
+    if (!has_back_edge) {
+      merge_block = terminator->getSuccessor(1);
```

Comment:
This is really subtle.  This is a strong dependence on the behaviour of the structurizer.
It says


    /// Control flow is expressed as a branch where the true exit goes into the
    /// "Then"/"Else" region, while the false exit skips the region

That's why you select the else branch here.  Needs a comment saying so (!)